algorithms/kernel/kmeans/kmeans_lloyd_helper.h

```diff
@@ -109,6 +110,47 @@ struct TlsTask
     size_t * cIndices         = nullptr;
 };
 
+template <Method method, typename algorithmFPType, CpuType cpu>
+struct BSHelper
+{
+    static size_t kmeansGetBlockSize(const size_t nRows, const size_t dim, const size_t clNum);
+};
+
+template <typename algorithmFPType, CpuType cpu>
+struct BSHelper<lloydDense, algorithmFPType, cpu>
+{
+    static size_t kmeansGetBlockSize(const size_t nRows, const size_t dim, const size_t clNum)
+    {
+        const double cacheFullness     = 0.8;
+        const size_t maxRowsPerBlock   = 512;
+        const size_t minRowsPerBlockL1 = 256;
+        const size_t minRowsPerBlockL2 = 8;
+        const size_t rowsFitL1         = (getL1CacheSize() / sizeof(algorithmFPType) - (clNum * dim)) / (clNum + dim) * cacheFullness;
+        const size_t rowsFitL2         = (getL2CacheSize() / sizeof(algorithmFPType) - (clNum * dim)) / (clNum + dim) * cacheFullness;
```

Comment:
I suppose it's applicable for Homogen mostly without data copying. Is it applicable for SOA, when we need to copy data?